jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/DemandingFlusher.java

```diff
@@ -0,0 +1,128 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.LongConsumer;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.CountingCallback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.websocket.core.Frame;
+
+public abstract class DemandingFlusher extends IteratingCallback implements DemandChain
+{
+    private final BiConsumer<Frame, Callback> _forwardFrame;
+    private final AtomicLong _demand = new AtomicLong();
+    private final AtomicReference<Throwable> _failure = new AtomicReference<>();
+    private LongConsumer _nextDemand;
+
+    private Frame _frame;
+    private Callback _callback;
+    private boolean _needContent = true;
+    private boolean _first = true;
+
+    public DemandingFlusher(BiConsumer<Frame, Callback> forwardFrame)
+    {
+        _forwardFrame = forwardFrame;
+    }
+
+    protected abstract boolean handle(Frame frame, Callback callback, boolean first);
+
+    @Override
+    public void demand(long n)
+    {
+        _demand.getAndUpdate(d -> Math.addExact(d, n));
+        iterate();
+    }
+
+    @Override
+    public void setNextDemand(LongConsumer nextDemand)
+    {
+        _nextDemand = nextDemand;
+    }
+
+    public void onFrame(Frame frame, Callback callback)
+    {
+        _frame = frame;
+        _callback = new CountingCallback(callback, 1);
```

Comment:
The implementation could multiple of its `payloadCallback`s at once. In the failure of this we need to fail the original frame callback, but we don't want to fail it multiple times. We can't rely on the state of the IteratingCallback because it may have another frame/callback at the time the `payloadCallback` is failed.

In addition to this there could be an error generated from inside the `IteratingCallback` which causes `onCompleteFailure` to be called, and we then want to fail the current frameCallback to relay that failure up to the `WebSocketCoreSession`, but the application may have access to this callback as well and might succeed/fail it at some point after.

So we do need this.